modules/indexer/code/bleve.go

```diff
@@ -89,10 +90,19 @@ func populateRepoIndexerAsynchronously() error {
 // should only be run when the indexer is created for the first time.
 func populateRepoIndexer(maxRepoID int64) {
 	log.Info("Populating the repo indexer with existing repositories")
+
+	isShutdown := graceful.GetManager().IsShutdown()
```

Comment:
The logic of the repo indexer will need some rewriting. If repo IDs 100 through 50 are indexed by this loop and Gitea is shut down, repos 49 through 1 will not be indexed at the next start (this function is only called if the index was empty).

We need a persistent queue to ensure the list of remaining repos is recovered at restart.

Reply:
AFAICS `populateRepoIndexer` is called at every startup and it crawls all of the repos

Reply:
Yup. We do.

Reply:
@guillep2k persistent queue of course. That's why I wrote levelqueue. It's for simple gitea instance. Maybe we should remove channel queue and make levelqueue or other persistent queue default. For bigger instances, we can use nsq/redis.

In fact, database will store the last indexed commit id. When next commit pushed, it will index from last commit to current commit and save current commit id to database. If Gitea didn't shutdown gracefully, it will lost all indexes which will be processed on channel queue until there are some new commit pushed to the default branch of that repositories.

Reply:
Don't worry I'm getting there...

Reply:
So getting this truly sorted will need to be in the next breakout of #8874

Reply:
I've done a partial fix - unindexed repos will be populated by populatereposindex. Dealing with the missing repo updates will need a queue I think.

Reply:
https://github.com/zeripath/gitea/commit/2ae0b41f4894bdab2121b753a4f35b55e97cc536